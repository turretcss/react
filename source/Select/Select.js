import React from 'react'
import classnames from 'classnames'

import { indicatorMap, sizeAliases } from '../constants'

const Select = ({
  options = [],
  empty = 'Please select ...',
  value = null,
  defaultValue = null,
  readOnly,
  size,
  color,
  className,
  onChange,
  children,
  ...props
}) => (
  <label
    className={classnames('select', className, {
      ['select-' + (sizeAliases[size] || size)]: size,
      [color]: indicatorMap[color],
    })}>
    <select
      {...props}
      value={toString(value)}
      onChange={event => !readOnly && onChange(children ? event.target.value : toValue(event.target.value, options))}>
      {children || renderOptions(options, empty, defaultValue)}
    </select>
  </label>
)

function renderOptions(options, empty, defaultValue) {
  const emptyOption = empty !== false && (
    <option key={0} value={toString(defaultValue)}>
      {empty}
    </option>
  )
  const valueOptions = options.map(({ label, value, ...props }, i) => (
    <option {...props} key={i + 1} value={value}>
      {label}
    </option>
  ))
  return emptyOption ? [emptyOption, ...valueOptions] : valueOptions
}

function toString(value) {
  return value === null ? '' : String(value)
}

function toValue(string, options) {
  return string === '' ? null : options.reduce((acc, { value }) => (String(value) === string ? value : acc), string)
}

export default Select
