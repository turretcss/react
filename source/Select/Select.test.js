import React from 'react'
import renderer from 'react-test-renderer'
import { renderToStaticMarkup } from 'react-dom/server'

import Select from './Select'

const options = [
  {
    value: 's',
    label: 'Small',
  },
  {
    value: 'm',
    label: 'Medium',
  },
  {
    value: 'l',
    label: 'Large',
  },
]

describe('Select', () => {
  it('renders', () => {
    const tree = renderer.create(<Select />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders array of options', () => {
    const tree = renderer.create(<Select options={options} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders array of options with "Small" option selected', () => {
    const tree = renderer.create(<Select value="s" options={options} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders custom empty option label', () => {
    const tree = renderer.create(<Select empty="All" options={options} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders without empty option', () => {
    const tree = renderer.create(<Select empty={false} options={options} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('spreads option attributes to options', () => {
    const options = [
      {
        value: 'one',
        label: 'One',
        disabled: true,
        'data-custom': 'yep',
      },
    ]
    const tree = renderer.create(<Select empty={false} options={options} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders children if specified', () => {
    const tree = renderer
      .create(
        <Select>
          <optgroup label="Group 1">
            <option value="1.1">Item 1.1</option>
            <option value="1.2">Item 1.2</option>
            <option value="1.3">Item 1.3</option>
          </optgroup>
          <optgroup label="Group 2">
            <option value="2.1">Item 2.1</option>
            <option value="2.2">Item 2.2</option>
            <option value="2.3">Item 2.3</option>
          </optgroup>
        </Select>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('spreads unknown props to select', () => {
    const tree = renderer.create(<Select name="thing" data-custom="on-select" />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('calls onChange when select changed', () => {
    let newValue = null
    const testInstance = renderer.create(<Select onChange={value => (newValue = value)} options={options} />).root
    testInstance.findByType('select').props.onChange({ target: { value: 's' } })
    expect(newValue).toEqual('s')
  })

  it('does not call onChange when readOnly', () => {
    let newValue = null
    const testInstance = renderer.create(<Select onChange={value => (newValue = value)} readOnly options={options} />)
      .root
    testInstance.findByType('select').props.onChange({ target: { value: 's' } })
    expect(newValue).toEqual(null)
  })

  it('respects value type when calling onChange', () => {
    let newValue = null
    const options = [{ value: 1, label: 'One' }]
    const testInstance = renderer.create(<Select onChange={value => (newValue = value)} options={options} />).root
    testInstance.findByType('select').props.onChange({ target: { value: '1' } })
    expect(newValue).toEqual(1)
  })

  it('converts empty string to null when calling onChange', () => {
    let newValue = false
    const options = [{ value: 1, label: 'One' }]
    const testInstance = renderer.create(<Select onChange={value => (newValue = value)} options={options} />).root
    testInstance.findByType('select').props.onChange({ target: { value: '' } })
    expect(newValue).toEqual(null)
  })

  it('does not respects value type when calling onChange if children supplied', () => {
    let newValue = null
    const testInstance = renderer.create(
      <Select onChange={value => (newValue = value)}>
        <option value={1}>One</option>
      </Select>
    ).root
    testInstance.findByType('select').props.onChange({ target: { value: '1' } })
    expect(newValue).toEqual('1')
  })
})
