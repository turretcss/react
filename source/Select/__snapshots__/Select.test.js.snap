// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders 1`] = `
<label
  className="select"
>
  <select
    onChange={[Function]}
    value=""
  >
    <option
      value=""
    >
      Please select ...
    </option>
  </select>
</label>
`;

exports[`Select renders array of options 1`] = `
<label
  className="select"
>
  <select
    onChange={[Function]}
    value=""
  >
    <option
      value=""
    >
      Please select ...
    </option>
    <option
      value="s"
    >
      Small
    </option>
    <option
      value="m"
    >
      Medium
    </option>
    <option
      value="l"
    >
      Large
    </option>
  </select>
</label>
`;

exports[`Select renders array of options with "Small" option selected 1`] = `
<label
  className="select"
>
  <select
    onChange={[Function]}
    value="s"
  >
    <option
      value=""
    >
      Please select ...
    </option>
    <option
      value="s"
    >
      Small
    </option>
    <option
      value="m"
    >
      Medium
    </option>
    <option
      value="l"
    >
      Large
    </option>
  </select>
</label>
`;

exports[`Select renders children if specified 1`] = `
<label
  className="select"
>
  <select
    onChange={[Function]}
    value=""
  >
    <optgroup
      label="Group 1"
    >
      <option
        value="1.1"
      >
        Item 1.1
      </option>
      <option
        value="1.2"
      >
        Item 1.2
      </option>
      <option
        value="1.3"
      >
        Item 1.3
      </option>
    </optgroup>
    <optgroup
      label="Group 2"
    >
      <option
        value="2.1"
      >
        Item 2.1
      </option>
      <option
        value="2.2"
      >
        Item 2.2
      </option>
      <option
        value="2.3"
      >
        Item 2.3
      </option>
    </optgroup>
  </select>
</label>
`;

exports[`Select renders custom empty option label 1`] = `
<label
  className="select"
>
  <select
    onChange={[Function]}
    value=""
  >
    <option
      value=""
    >
      All
    </option>
    <option
      value="s"
    >
      Small
    </option>
    <option
      value="m"
    >
      Medium
    </option>
    <option
      value="l"
    >
      Large
    </option>
  </select>
</label>
`;

exports[`Select renders without empty option 1`] = `
<label
  className="select"
>
  <select
    onChange={[Function]}
    value=""
  >
    <option
      value="s"
    >
      Small
    </option>
    <option
      value="m"
    >
      Medium
    </option>
    <option
      value="l"
    >
      Large
    </option>
  </select>
</label>
`;

exports[`Select spreads option attributes to options 1`] = `
<label
  className="select"
>
  <select
    onChange={[Function]}
    value=""
  >
    <option
      data-custom="yep"
      disabled={true}
      value="one"
    >
      One
    </option>
  </select>
</label>
`;

exports[`Select spreads unknown props to select 1`] = `
<label
  className="select"
>
  <select
    data-custom="on-select"
    name="thing"
    onChange={[Function]}
    value=""
  >
    <option
      value=""
    >
      Please select ...
    </option>
  </select>
</label>
`;
